### PR TITLE
Triplet Losses - second chance

### PR DESCRIPTION
**Pull request by TuSKan**
Hi Jan,

This is a new PR about triplet losses, completely refactored, separated in a _triplet.go_ and _triplet_test.go_ files,  including all your recommendations.
I have implemented the triplet loss simple version (average on all pos and negs), and the hard (average on hard pos and hard negs) and semi-hard (average on pos and negs on the margin) variations. My semi-hard implementation is slight different from tensorflow implementation, i believe it's more intuitive, it was difficult to figure out how to do it.
I have include also a additional option to soft the loss, rather than take the _max(loss,0)_.
The documentation is completed and following [go docs](https://go.dev/doc/comment) with all relevant references included.

Please, let me know what do you think about this code, I hope it's not creep anymore.

Any comments and suggestions are welcome.

(obs.: I need to reset the original branch because some mistakes that I make during merge process)

### CODE REVIEW

File: graph/gen_backend_ops.go
Comment (janpfeifer) [inline graph/gen_backend_ops.go:1]: In this case these changes will be overwritten later.

All files starting with `gen_` were automatically generated from other files (search for `//go:generate` tags in this directory).

If you can, just reset this file to the original value (or not), I'll fix it later on the source and regenerate this file.

File: ml/train/losses/triplet.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
+/*
+ *	Copyright 2023 Jan Pfeifer
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:2]: I'm no lawyer, but I think you should use your name here 😄 As long as the project remains open-source (Apache 2 License in this case).

File: ml/train/losses/triplet.go [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+)
+
+type TripletLossDistance int
+
+const (
+	TripletLossDistanceL2 TripletLossDistance = iota
+	TripletLossDistanceSquaredL2
+	TripletLossDistanceCosineSimilarity
+)
+
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:30]: What about adding the following line below:

//go:generate enumer -type=TripletLossDistance -trimprefix=TripletLossDistance -transform=snake -values -text -json -yaml triplet.go

And then go to the losses directory and run: `go generate .`

File: ml/train/losses/triplet.go [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+// pairwiseL2Distances Compute the 2D matrix of L2/Squared L2 distances between all the embeddings.
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:31]: nit: s/Compute/computes

File: ml/train/losses/triplet.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+//
+// Returns:
+//   - *Node 2-D tensor of shape (batch_size, batch_size)
+func pairwiseL2Distances(embeddings *Node, squared bool) *Node {
+	g := embeddings.Graph()
+	batchSize := embeddings.Shape().Dim(0)
+	dtype := embeddings.DType()
+	// ||a - b||^2 = ||a||^2  - 2 <a, b> + ||b||^2
+	// Get the dot product between all embeddings
+	// shape (batch_size, batch_size)
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:46]: Another nit picking: in GoMLX I've been using asserts instead of comments on the shape.
Because it serves both as a comment, and as a runtime check:

```
dotProduct.AssertDims(batchSize, batchSize)
```

File: ml/train/losses/triplet.go [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+	// Get squared L2 norm for each embedding. We can just take the diagonal of `dot_product`.
+	// This also provides more numerical stability (the diagonal of the result will be exactly 0).
+	// shape (batch_size,)
+	squareNorm := MaskedReduceSum(dotProduct, Diagonal(g, batchSize), 0)
+
+	// Compute the pairwise distance matrix as we have:
+	// ||a - b||^2 = ||a||^2  - 2 <a, b> + ||b||^2
+	// shape (batch_size, batch_size)
+	distances := Add(Add(
+		ExpandDims(squareNorm, 1),
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:58]: ExpandDims is actually deprecated -- in favor of `ExpandAxes` or `InsertAxes` depending how you interpret the axes being given (before or after being inserted). 

In this case you want `ExpandAxes`, and the exact same parameters.

File: ml/train/losses/triplet.go [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+		MulScalar(dotProduct, -2.0)),
+		ExpandDims(squareNorm, 0))
+
+	// Because of computation errors, some distances might be negative so we put everything >= 0.0
+	distances = MaxScalar(distances, 0.0)
+
+	if !squared {
+		// Because the gradient of sqrt is infinite when distances == 0.0 (ex: on the diagonal)
+		// we need to add a small epsilon where distances == 0.0
+		zero := Scalar(g, dtype, 0.0)
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:68]: `zero := ScalarZero(g, dtype)`

File: ml/train/losses/triplet.go [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+		eps := epsilonForDType(g, dtype)
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:69]: You don't actually need epsilon here. It could be anything that yields non-NaN values for SQRT.
(it's nice to avoid epsilon specially in low precision types like f16, bf16)

What about using `one` instead ?  The epsilon is just a placeholder here.

```
one := ScalarOne(g, dtype)
mask := LessOrEqual(distances, zero)   // Just in case of numerical issues that lead to small negative distances.
distances = Where(mask, one, distances)
distances = Sqrt(distances)
distances = Where(mask, zero, distances)

File: ml/train/losses/triplet.go [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+		distances = Sqrt(distances)
+
+		// Correct the epsilon added: set the distances on the mask to be exactly 0.0
+		distances = Where(mask, zero, distances)
+	}
+
+	return distances
+}
+
+// pairwiseCosineDistances  Compute the 2D matrix of cosine distances between all the embeddings.
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:82]: nit: s/Compute/computes

File: ml/train/losses/triplet.go [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+	// normalize input
+	embeddingsL2Normalize := L2Normalize(embeddings, 1)
+	// create adjacent matrix of cosine similarity
+	distances := OneMinus(MatMul(embeddingsL2Normalize, Transpose(embeddingsL2Normalize, 0, 1)))
+	// ensure all distances >= 0.0
+	distances = MaxScalar(distances, 0.0)
+	return distances
+}
+
+// validTripletMask Return a 3D mask where mask[a, p, n] is True iff the triplet (a, p, n) is valid
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:99]: s/Return/returns

"... where a=anchor, p=positive and n=negative examples."

File: ml/train/losses/triplet.go [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+//
+// A triplet (i, j, k) is valid if:
+//   - i, j, k are distinct
+//   - labels[i] == labels[j] and labels[i] != labels[k]
+//
+// Parameters:
+//   - labels *Node labels of the batch, of size (batch_size,)
+//
+// Returns:
+//   - *Node a mask of valid triplet(a, p, n) with shape (batch_size,batch_size,batch_size,batch_size)
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:109]: I think there one too many `batch_size` here ? There should be only 3, no ?

File: ml/train/losses/triplet.go [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+	//  Check that i, j and k are distinct
+	indices := LogicalNot(DiagonalWithValue(Const(g, true), batchSize))
+	iNotEqualj := InsertAxes(indices, 2)
+	iNotEqualk := InsertAxes(indices, 1)
+	jNotEqualk := InsertAxes(indices, 0)
+
+	distinct := And(And(iNotEqualj, iNotEqualk), jNotEqualk)
+
+	// Check if labels[i] == labels[j] and labels[i] != labels[k]
+	equal := Squeeze(Equal(InsertAxes(labels, 0), ExpandDims(labels, 1)))
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:122]: `ExpandAxes`, `InsertAxes` and `ExpandDims` are equivalent, but `ExpandDims` is deprecated.

Ideally, I was going to suggest to pick `ExpandAxes` or `InsertAxes` and only use this one everywhere. Whichever you like more.

Some explanation (sry if you already know):

- `ExpandAxes` the axes given are considered from the point of view of the expanded tensor.
- `InsertAxes` the axes given are considered the positions in the original tensor, before which to insert axes.

Eg.: to insert 3 new axes (of dimension 1) at the start of a tensor x one could do (they are equivalent):

* InsertAxes(x, 0, 0, 0)  // Insert all 3 before position 0 of the original x.
* ExpandAxes(x, 0, 1, 2)  // The new axes will be in position 0, 1, 2 in the new expanded tensor.

File: ml/train/losses/triplet.go [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+	iEqualj := InsertAxes(equal, 2)
+	iEqualk := InsertAxes(equal, 1)
+
+	valid := And(iEqualj, LogicalNot(iEqualk))
+
+	return And(distinct, valid)
+}
+
+// TripletLoss Compute the triplet loss of all the valid triplet over a batch of embeddings.
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:131]: nit:
* s/Compute/computes
* s/valid triplet/valid triplets
*

File: ml/train/losses/triplet.go [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+//
+// # We generate all the valid triplets and average the loss over the positive ones.
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:133]: nit: remove the "#"?

File: ml/train/losses/triplet.go [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+	// (where label(a) != label(p) or label(n) == label(a) or a == p)
+	// shape (batch_size,batch_size,batch_size,batch_size)
+	valid := validTripletMask(labels0)
+
+	tripletLoss = Where(valid, tripletLoss, zero)
+
+	// Remove negative losses (i.e. the easy triplets)
+	tripletLoss = MaxScalar(tripletLoss, 0.0)
+
+	// Count number of positive triplets (where loss > 0)
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:187]: There is something that does this selective mean better in GoMLX. I think you can replace the lines below with:

lossMask := GreaterThan(tripletLoss, eps)   // I'm not sure about eps here, couldn't it be simply >= zero, as the comment above says ?
loss := MaskedReduceMean(tripletLoss, lossMask) 

See https://github.com/gomlx/gomlx/blob/main/graph/ops.go#L847

File: ml/train/losses/triplet.go [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+	labels0 := labels[0]
+	weights, mask := CheckLabelsForWeightsAndMask(labels0.Shape(), labels)
+
+	g := predictions0.Graph()
+	dtype := predictions0.DType()
+	eps := epsilonForDType(g, dtype)
+	zero := ScalarZero(g, dtype)
+	batchSize := labels0.Shape().Dim(0)
+
+	// shape (batch_size, batch_size)
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:315]: Optional: What about refactoring the block below repeated in the the function as:

```go
func pairwiseDistance(metric TripletLossDistance, embeddings *Node) *Node {
	var distances *Node
	switch distanceMetric {
	case TripletLossDistanceL2:
		distances = pairwiseL2Distances(predictions0, false)
	case TripletLossDistanceSquaredL2:
		distances = pairwiseL2Distances(predictions0, true)
	case TripletLossDistanceCosineSimilarity:
		distances = pairwiseCosineDistances(predictions0)
	}
	return distance   
}
```

?

File: ml/train/losses/triplet.go [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+		distances = pairwiseL2Distances(predictions0, true)
+	case TripletLossDistanceCosineSimilarity:
+		distances = pairwiseCosineDistances(predictions0)
+	}
+
+	// shape (batch_size, batch_size)
+	indicesNotEqual := LogicalNot(DiagonalWithValue(Const(g, true), batchSize))
+
+	// shape (batch_size, batch_size)
+	labelsEqual := Squeeze(Equal(InsertAxes(labels0, 0), ExpandDims(labels0, 1)))
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:330]: Same as above: either `InsertAxes` or `ExpandAxes`, whichever is your favorite.

File: ml/train/losses/triplet.go [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+	// shape (batch_size, batch_size)
+	negativesMask := LogicalNot(labelsEqual)
+
+	// First, we need to get a mask for every valid positive (they should have same label)
+	// shape (batch_size, batch_size)
+	positivesMask := And(indicesNotEqual, labelsEqual)
+
+	// largest d(a, n)
+	// shape (batch_size, 1)
+	axisMinimums := ReduceAndKeep(distances, ReduceMin, 1)
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:342]: I think these 2 lines could be replaced by the simpler (and likely faster):

negativesInside := ExpandAxes(MaskedReduceMax(distances, negativesMask, 1), -1)

File: ml/train/losses/triplet.go [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+	negativesInside := Add(ReduceAndKeep(
+		Where(negativesMask, Sub(distances, axisMinimums), ZerosLike(distances)),
+		ReduceMax, 1), axisMinimums)
+
+	// d(a, n) > d(a, p)
+	// axis 1 are the negatives
+	// shape (batch_size, batch_size, batch_size)
+	outsideMask := Transpose(And(
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:350]: Sry, I'm not following why the transpose here. Can't this be achieved by just inserting the axes at different positions ? 

Maybe split the lines ? 

```
outsideMask := GreaterThan(InsertAxes(distances, 1), InsertAxes(distances, 0))
outsideMask = And(outsideMask, InsertAxes(negativesMask,1))
outsideMask = Transpose(validPosNegDistances, 0, 1)
```

But still, I have the feeling the transpose shouldn't be needed, if the axes are inserted right ... No need to do anything, just if you feel like solving this 😄 I may be wrong also ...

File: ml/train/losses/triplet.go [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+// TripletHardLoss Compute the triplet loss using the hardest positive and the hardest negative for each anchor over a batch of embeddings.
+//
+//	We select the hardest positive and the hardest negative for each anchor
+//
+// Parameters:
+//   - labels *Node labels of the batch, of size (batch_size,)
+//   - embeddings *Node 2-D Tensor of shape (batch_size, embed_dim)
+//   - margin float64 margin for triplet loss
+//   - distanceMetric TripletLossDistanceMetric Metric(L2, squared L2 or cosine similarly) to compute the distance matrix
+//
```
Comment (janpfeifer) [inline ml/train/losses/triplet.go:216]: Missing "soft" parameter description.

Also, shouldn't "soft" be available for the simple TripletLost above ?

**Comment by TuSKan** [comment]
I just found a very elegant implementation of triplet loss on tensorflow similarly package [triplet_loss.py](https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/losses/triplet_loss.py) that follow the same approach that I did on semi-hard negatives. 
Do you think it worth refactor the code to have a similar interface with one function, options on pos e negs sampling, more modular ?

**Review by janpfeifer — COMMENTED** [review]
First let me say the code looks very good!

Hence, I wanted to spend time doing a nice code review. I haven't completely finished yet, but I didn't want to block you so I'll send the comments I already have.

Many are small GoMLX different (hopefully simpler) way of doing things, other a simple typos.

looks great, thanks!

**Comment by janpfeifer** [comment]
> I just found a very elegant implementation of triplet loss on tensorflow similarly package [triplet_loss.py](https://github.com/tensorflow/similarity/blob/master/tensorflow_similarity/losses/triplet_loss.py) that follow the same approach that I did on semi-hard negatives.
> Do you think it worth refactor the code to have a similar interface with one function, options on pos e negs sampling, more modular ?

That also sounds awesome. I find it simpler to have one interface to TripletLoss with options (as opposed to many functions). But it's a choice thing, one can also have several functions. In GoMLX I've been doing the pattern of having one function with many optional parameters, using the "builder pattern" --> e.g: (1) the simpler [`layers.Convolution`](https://github.com/gomlx/gomlx/blob/main/ml/layers/convolution.go#L63) function; (2) the [`layers.fnn` package](https://github.com/gomlx/gomlx/blob/main/ml/layers/fnn/fnn.go); the [`layers.kan` package](https://github.com/gomlx/gomlx/tree/main/ml/layers/kan). I use this pattern everywhere.

If you have no strong preference/reason to do different, consider using it for consistency.

Also ... I mention previously, if someone is using the official trainer, it uses a [`losses.LossFn`](https://github.com/gomlx/gomlx/blob/main/ml/train/losses/losses.go#L44) as an input parameter. Consider writing a function that returns a LossFn closure, so it can be passed to the trainer. See [losses.MakeAdaptivePowerLoss()] -- this one doesn't take optional parameters, so it doesn't use the "builder pattern".

Sorry that was a lot, I hope it makes sense to you. But let me know if you have any questions, or don't agree with something. We can also schedule a VC to chat if it is simpler.

cheers
